### ISSUE TITLE
False error TS4094: ... exported class expression may not be private or protected.

### ISSUE THREAD

**Issue by trusktr** [microsoft/TypeScript#36060]
Re-opening of #30355

**Comment by RyanCavanaugh** [comment]
Wouldn't it be cool if this had some text in it?

**Comment by trusktr** [comment]
It doesn't need, the other issue has details, but it is closed. I just think that this is a valid problem that is faced much too often by TS devs.

**Comment by web-padawan** [comment]
Related to #17293 (actually the same) - see https://github.com/microsoft/TypeScript/issues/17293#issuecomment-522491710 

Related to #17744 - see my comment explaining why this is important https://github.com/microsoft/TypeScript/issues/17744#issuecomment-558990381

**Comment by RonnieLincoln** [comment]
any process?

Encountered this issure when trying Builder Pattern.

**Comment by miguel-leon** [comment]
Couldn't typescript just generate the class type (like with any other class), put that type before the function, and make the function return that type? instead of trying to inline the return type of the function (where the problem is).

This seems imperative for the Mixin pattern.

**Comment by MichaelMitchell-at** [comment]
This seems like it would be a way to annotate classes that have computed properties when using isolated declarations. Right now there is no way I'm aware of to enable isolated declarations for code of this form without loss of type information. 
```ts
export class Foo implements Iterable<number> {
  [Symbol.iterator]: () => {
  // ^ Computed property names on class or object literals cannot be inferred with --isolatedDeclarations.ts(9038)
    ...
  }
}
```

If anonymous class types were allowed then we could perhaps do something like
```ts
export const Foo: class implements Iterable<number> {} = class {
  [Symbol.iterator]: () => {
    ...
  }
}
```

cc @dragomirtitian @weswigham

**Comment by trusktr** [comment]
I thought I'd try to be a little bit cool, so I added some text for some context. 😎😄

A notable use case for this is support for class-factory mixins.

This code,

```ts
type Constructor<T = object, A extends any[] = any[], Static = {}> = (new (...a: A) => T) & Static

function CoolMixin(Base: Constructor<HTMLElement>) {
  return class CoolClass extends Base {
    private foo = 123

    meth(other: CoolClass) {
      console.log(other.foo)
    }
  }
}
```

currently produces the type error

```
Property 'foo' of exported anonymous class type may not be private or protected. ts(4094)
```

[playground](https://www.typescriptlang.org/play/?isolatedModules=true&verbatimModuleSyntax=true#code/C4TwDgpgBAwg9gOwM7AE4FcDGw6oDwAqUAvFHAEYBWE2ANFAIJQQAewECAJklAIYIgA2gF0SfASPoBlYL2ABLTGIDeAXwB8YgBQIIAdyhaAdCd4AuRgEoSmgtYBkUGXMUAoVwDN0CbPMSw4OAAbAFl5FnkELQAhXiQIC3hkNCwcfAAJAhCAGQBRIIgAWw5gdWtlVygoVAhgdFQEKEwguJ54YJgWpB5Wdi4eWPioCqqqsFR5ADc5aA9AsQBGACYAZndRqGLgAAstOB2IVETAoM7W8sqNpsQkYIgjILgAcz2D1CM5uEtLqtVLv9UQA)

Note that using the newer `#hashtag` private syntax,

```ts
type Constructor<T = object, A extends any[] = any[], Static = {}> = (new (...a: A) => T) & Static

function CoolMixin(Base: Constructor<HTMLElement>) {
  return class CoolClass extends Base {
    #foo = 123

    meth(other: CoolClass) {
      console.log(other.#foo)
    }
  }
}
```

does not have a type error!

[playground](https://www.typescriptlang.org/play/?isolatedModules=true&verbatimModuleSyntax=true#code/C4TwDgpgBAwg9gOwM7AE4FcDGw6oDwAqUAvFHAEYBWE2ANFAIJQQAewECAJklAIYIgA2gF0SfASPoBlYL2ABLTGIDeAXwB8YgBQIIAdyhaAdCd4AuRgEoSmgtYBkUGXMUAoVwDN0CbPMSw4OAAbAFl5FnkELQAhXiQIC3hkNCwcfAAJAhCAGQBRIIgAWw5gdWtlVygoVAhgdFQEKEwguJ54YJgWpB5Wdi4eWPioCqqqgGIPQLEARgAmAGZ3Uahi4AALLTh1iFREwKDO1vLK5abEJGCIIyC4AHNN7dQjCcDLE6rVE8-VIA)

However! Mixins that use `#hashtag` private fields will experience runtime errors in JavaScript. This code,

```ts
type Constructor<T = object, A extends any[] = any[], Static = {}> = (new (...a: A) => T) & Static

function CoolMixin(Base: Constructor<HTMLElement>) {
  return class CoolClass extends Base {
    #foo = 123

    meth(other: CoolClass) {
      console.log(other.#foo)
    }
  }
}

class One extends CoolMixin(HTMLElement) {}
customElements.define('x-one', One)

const one = new One()

class Two extends CoolMixin(HTMLElement) {}
customElements.define('x-two', Two)

const two = new Two

two.meth(one) // Runtime error!
```

throws the error

```
TypeError: Cannot read private member #foo from an object whose class did not declare it
```

[playground](https://www.typescriptlang.org/play/?target=99&isolatedModules=true&verbatimModuleSyntax=true#code/C4TwDgpgBAwg9gOwM7AE4FcDGw6oDwAqUAvFHAEYBWE2ANFAIJQQAewECAJklAIYIgA2gF0SfASPoBlYL2ABLTGIDeAXwB8YgBQIIAdyhaAdCd4AuRgEoSmgtYBkUGXMUAoVwDN0CbPMSw4OAAbAFl5FnkELQAhXiQIC3hkNCwcfAAJAhCAGQBRIIgAWw5gdWtlVygoVAhgdFQEKEwguJ54YJgWpB5Wdi4eWPioCqqqgGIPQLEARgAmAGZ3Uahi4AALLTh1iFREwKDO1vLK5abEJGCIIyC4AHNN7dQjCcDLE6rVE8-P12bWqAA8rpmGwONwAsEwhEopkcvkiiVyp9MOgUHBCvDighgEgjJwIB5IhAtAByFgAWkQEBJ9CBEDev3OwDIwNIugMdK0DL+3SgBD0cBBfXB7VC4UiWlheQKWOASN+qJwGJlJVx+MJulJFOAAppfIF3KZUB1grZ+n1cHcJqMqw2VOsAHoHVAAEreBTFZioVC4ACEQA)

The simplest workaround is to use a convention like prefixing members with `_` (one underscore) for protected, and prefixing members with `__`  (two underscores) for private, abandoning `protected`, `private`, and `#private` altogether.

This works:

```ts
type Constructor<T = object, A extends any[] = any[], Static = {}> = (new (...a: A) => T) & Static

/**
 * Members prefixed with __ are private, don't touch or you're a gonner!
 */
function CoolMixin(Base: Constructor<HTMLElement>) {
  return class CoolClass extends Base {
    __foo = 123

    meth(other: CoolClass) {
      console.log(other.__foo)
    }
  }
}

class One extends CoolMixin(HTMLElement) {}
customElements.define('x-one', One)

const one = new One()

class Two extends CoolMixin(HTMLElement) {}
customElements.define('x-two', Two)

const two = new Two

two.meth(one)
```

[playground](https://www.typescriptlang.org/play/?target=99&isolatedModules=true&verbatimModuleSyntax=true#code/C4TwDgpgBAwg9gOwM7AE4FcDGw6oDwAqUAvFHAEYBWE2ANFAIJQQAewECAJklAIYIgA2gF0SfASPoBlYL2ABLTGIDeAXwB8YgBQIIAdyhaAdCd4AuRgEoSmgtYBkUGXMUAoVwHoAVF9dQvUACyEAC25BCoPGCoEABm8iwQnFB68sAAFlAA+ll8MVDR8gBuchD0nIgA5MBQOFiZuFAgcOiVADZtUOFQ8TGcAIR+Xh6usegI2PKIsHBwbYEJ8ghaAEK8SBAW8MhoWDj4ABIEgQAyAKJtoRzA6tbKflAxwOioCFCYbes88HMwn0g8VjsLg8NYbKD3KBQ7JZWKzMQARgATABmdzQqAhCAZLRwDIRLazNp-L53B4YzCIJBzCBGNpwADmuPxqCMOThcEs5KgqgevN5rg+XygAHldMw2BxuDM5gsWEstEdThcrghgHdeZh0Cg4CEVVi1UgjJw4ksIFpKiwALSICCVehiiBcwVUmq2sS6AyOrTOoUAqAEPRwCXA6U-eaLZZK86XA3qiGa7U4PWx65Gk3xXQW63AIP2gNB32u2pBj36Atwdy5uBGLE422WIA)

For now, I use underscore conventions because that's easy! It does mean that the "protected" and "private" members are technically public, visible for IDE autocompletion in contexts when they otherwise wouldn't be, which is not ideal, but the visible underscores should be enough of a clue, especially if the convention is well documented.

Another workaround, for private members, it to use WeakMaps, but that's a bit more cumbersome than the simple underscore convention.

**Comment by MichaelMitchell-at** [comment]
Isolated declarations was later extended to support the case I brought up. My previous comment can be disregarded.
